packages/common/src/TenderlySimulation.ts

```diff
@@ -0,0 +1,216 @@
+  const body = createRequestBody(simulationParams);
+  const headers = { "X-Access-Key": tenderlyEnv.apiKey };
+
+  // Send Tenderly simulation API request with retries.
+  const response = await retry(async () => {
+    const fetchResponse = await fetch(url, {
+      method: "POST",
+      headers: headers,
+      body: JSON.stringify(body),
+    });
+    if (!fetchResponse.ok) {
+      throw new Error(`Simulation API returned HTTP ${fetchResponse.status}: ${fetchResponse.statusText}`);
+    }
+    return fetchResponse;
+  }, retryOptions);
+
+  // If the HTTP response was OK, we expect the response body should be a JSON object containing expected Tenderly
+  // simulation response properties.
+  const apiResponse = await response.json();
+  if (!isTenderlyAPIResponse(apiResponse)) {
+    throw new Error(`Failed to parse Tenderly simulation API response: ${JSON.stringify(apiResponse)}`);
+  }
+  return apiResponse;
+};
+
+const isProjectPublic = async (tenderlyEnv: TenderlyEnvironment): Promise<boolean> => {
+  const url = `https://api.tenderly.co/api/v1/public/account/${tenderlyEnv.user}/project/${tenderlyEnv.project}`;
+
+  // Return true only if the project API responds OK and the project is public. On any error, return false.
+  try {
+    const response = await fetch(url);
+    if (!response.ok) throw new Error(`Project API returned HTTP ${response.status}: ${response.statusText}`);
+    const projectResponse = (await response.json()) as { project: { public: boolean } };
+    return projectResponse.project.public;
+  } catch {
+    return false;
+  }
+};
+
+const getResultUrl = async (
+  simulationId: string,
+  tenderlyEnv: TenderlyEnvironment,
+  fork?: ForkParams
+): Promise<ResultUrl> => {
+  const publicUrl = `https://dashboard.tenderly.co/public/${tenderlyEnv.user}/${tenderlyEnv.project}/${
+    fork !== undefined ? "fork-simulation" : "simulator"
+  }/${simulationId}`;
+  const privateUrl = `https://dashboard.tenderly.co/${tenderlyEnv.user}/${tenderlyEnv.project}/${
+    fork !== undefined ? "fork/" + fork.id + "/simulation" : "simulator"
+  }/${simulationId}`;
+
+  return (await isProjectPublic(tenderlyEnv)) ? { url: publicUrl, public: true } : { url: privateUrl, public: false };
+};
+
+export const simulateTenderlyTx = async (
+  simulationParams: TenderlySimulationParams,
+  retryOptions: RetryOptions = { retries: 0 } // By default, do not retry, but the caller can override this.
+): Promise<TenderlySimulationResult> => {
+  // Will throw if required environment variables are not set.
+  const tenderlyEnv = processEnvironment();
+
+  // Will throw if simulation parameters are invalid.
+  validateSimulationParams(simulationParams);
+
+  // Will throw if Tenderly API request fails or returns unparsable response.
+  const simulationResponse = await getSimulationResponse(simulationParams, tenderlyEnv, retryOptions);
+
+  // Get the URL to the simulation result page. If project is not public, the URL will be private (requires login).
```

Comment:
Just making sure I understand, private simulations would require login and access to the project, right? Or would anyone logging in to any account suffice (I assume not)?

If it requires project access, what do you see as the use case for producing private simulations here? Just added flexibility or do you see us/others using this feature?

Reply:
Yes, private simulation links require login and access to the project. We pass the public flag in ResultUrl to the consuming application so that it can decide if it is ok with just a private link. This might be acceptable to on-call team sharing access to a private project where monitor bot produces private link, while UI application could require this to always being public and handle this.